### PR TITLE
Update load_prepared_dataset_in_dataloaders.py

### PR DESCRIPTION
**Pull request by YuzhouLin**
To improve the code readability and reduce the code redundancy.

1. It is found 'get_dataloader' is a key function here, which has been used a lot, and the difference between the function 'load_dataloaders' and 'load_dataloaders_for_training_without_TL' is only the data aggregation. Therefore, the idea is to add the data aggregation function into the 
'get_dataloader'.

2. To improve the way to get validation dataset by allowing the 'get_dataloader' to take cycle_used (a list) as an input.

3. There is no need to use conditions for returning different outputs since you know which one you need when you are using the function.
For example, if you know you only need train and validation sets by setting the 'get_test_set' as False, you could simply use the function 
'load_dataloaders' as follows:
dataloaders_train, dataloaders_validation, _= load_dataloaders()

4. If these modifications seem reasonable to you, similar changes can be done for load_dataloaders_for_regression and get_dataloader_for_regression and
on those files which apply 'load_dataloaders' or 'load_dataloaders_for_training_without_TL', a little modification on the way to call functions needs to be done as well.

Please let me know if there is something unclear or a discussion is needed.

### CODE REVIEW

**Review by UlysseCoteAllard — APPROVED** [review]
Hi, 
First, I apologize for answering so late, I'm in a rush for a paper's deadline. Second, thank you very much, your changes definitively improve the code. Unfortunately, I will not have time to make these changes for the other files in the short term. 

By curiosity, can I ask what you are working on, it's always great to learn about EMG-related research! 

Have a great day!

**Comment by YuzhouLin** [comment]
Hi, Ulysse,

Thanks for your approving and your interests. I believe my research interests highly match yours, and this was the main reason why I was one of your followers since long time ago.

So in general, we are working on the sEMG-based hand gesture recognition and hope the investigation of such research question would be helpful for amputees who could precisely control the robotic arms in practical. However, due to some limitations such as lacking of expensive mechanical devices in our department, my research has a specific focus on the algorithms for the hand gesture classification.

Previously, I have tried to investigate muscle connectivity when subjects are performing a hand gesture. It was only a small attempt since muscle connectivity is a quite complex research question. My another piece of work was to solve the inter-subject hand gesture recognition problem. We proposed a simple normalisation approach and the performance was compared with the transfer learning you proposed. This is also currently we are working on. We tried to extend it to a journal paper by comparing it against the most recent work done by Evan who investigated the potential of using ADANN on the inter-subject hand gesture cognition by allowing one cycle data from the end-user for adaptation. Maybe no need to say too much about it since I found this paper by your recommendation somewhere.

Another important project I have done and submitted recently was about analysing the reliability of such hand gesture classifier. I hope you would agree with me about the following finding: the current literature research on this filed only cover the effectiveneness and robustness of hand gesture recognition with sEMG. In short, the effectiveness refers to the accuracy and the robustness is proposed to answer the question: if the high accuracy could be remained on different scenarios. However, the reliability refers to a different question: if the classifier is reliable? The difficulty comes up with how to analyse and ‘define’ such reliability. We first analyse it from the perspective of model uncertainty. Then the model reliability can be evaluated by the quality of the model uncertainty.

I hope this is clear and could give you a brief idea of what kinds of research questions we have been working on. I am glad to share more if you are interested in. Please do not hesitate to let me know if you have any questions.


Sincerely
Yuzhou

发件人: ***@***.***>
发送时间: 01 July 2021 20:43
收件人: ***@***.***>
抄送: Yuzhou ***@***.***>; ***@***.***>
主题: Re: [UlysseCoteAllard/sEMG_handCraftedVsLearnedFeatures] Update load_prepared_dataset_in_dataloaders.py (#2)

CAUTION: This email originated from outside of the organisation. Do not click links or open attachments unless you recognise the sender and know the content is safe.


@UlysseCoteAllard approved this pull request.

Hi,
First, I apologize for answering so late, I'm in a rush for a paper's deadline. Second, thank you very much, your changes definitively improve the code. Unfortunately, I will not have time to make these changes for the other files in the short term.

By curiosity, can I ask what you are working on, it's always great to learn about EMG-related research!

Have a great day!

―

**Comment by UlysseCoteAllard** [comment]
Thank you for this overview, very interesting and I like your out-of-the-box approach! 

I remember reading some time ago your paper "A normalisation approach improves the performance of inter-subject sEMG-based hand gesture recognition with a ConvNet" which I found quite interesting, looking forward to seeing the journal paper version of it! I wonder if it would be possible to apply your normalization in an unsupervised way (i.e. without needing to explicitly know the different gestures being performed). For example, learning a user-independent classifier from an algorithm like ADANN and then pseudo-labelling the recording from the new user (using a pseudo-labelling scheme such as the one described in "Unsupervised domain adversarial self-calibration for electromyography-based gesture recognition"). If the pseudo-labelling is not too noisy, your normalization might then be used to great effect!

Good luck with the paper on reliability, I'll be looking forward to reading it! 

Have a great day :)
